File: 0700-0799/0701. Insert into a Binary Search Tree/0701. Insert into a Binary Search Tree.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @param {number} val
 * @return {TreeNode}
 */
const insertIntoBST = function (root, val) {
    if (root === null) {
        return new TreeNode(val)
    } else {
        val < root.val ? root.left = insertIntoBST(root.left, val) : root.right = insertIntoBST(root.right, val);
    }
    return root;
};
const insertIntoBST2 = function (root, val) {
    if (root === null) return new TreeNode(val);
    let cur = root;
    while (true) {
        if (cur.val <= val) {
            if (cur.right !== null) cur = cur.right;
            else {
                cur.right = new TreeNode(val);
                break;
            }
        } else {
            if (cur.left !== null) cur = cur.left;
            else {
                cur.left = new TreeNode(val);
                break;
            }
        }
    }
    return root;
};

File: 0700-0799/0701. Insert into a Binary Search Tree/0701. Insert into a Binary Search Tree.py
# Definition for a binary tree node.
class TreeNode:

    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


from typing import Optional


class Solution:

    def insertIntoBST(self, root: Optional[TreeNode], val: int) -> Optional[TreeNode]:
        if root is None:
            return TreeNode(val)
        else:
            if val < root.val:
                root.left = self.insertIntoBST(root.left, val)
            else:
                root.right = self.insertIntoBST(root.right, val)
        return root

    def insertIntoBST2(self, root: Optional[TreeNode], val: int) -> Optional[TreeNode]:
        if root is None:
            return TreeNode(val)
        cur = root
        pre = None
        while cur is not None:
            pre = cur
            cur = cur.right if val > cur.val else cur.left
        if val > pre.val:
            pre.right = TreeNode(val)
        else:
            pre.left = TreeNode(val)
        return root
